### ISSUE TITLE
[clang][analyzer] False positive: non-inlined recursive call invalidates an unchanged pointer

### ISSUE THREAD

**Issue by bozicrHT** [llvm/llvm-project#222313]
**Description:**
A proven non-null var constraint is lost across a call to a static recursive function.

**Reproducer:**
```c
void clang_analyzer_eval(int);
void clang_analyzer_explain(void *);

static void a(int **p, unsigned n)
{
    if (n)
        a(p, n-1);
}

int entry(unsigned n)
{
    int x = 0;
    int *p = &x;

    a(&p, n);
    clang_analyzer_eval(p != 0); // TRUE and FALSE, expected only TRUE
    clang_analyzer_explain(p); // symbol of type 'int *' conjured at CFG element

    return *(p ? p : 0); // core.NullDereference
}
```

**Comment by NagyDonat** [comment]
This invalidation is justified in the framework followed by the analyzer. The analyzer only "sees" the single execution path that is being followed, and – in general – cannot prove that the recursive call doesn't contain e.g. `*p = NULL`.

Of course in the concrete example in the bug report a human can easily see that `f()` never assigns anything to `*p`, but we would need to introduce a specific AST inspection logic if we wanted to "teach" this to the analyzer, and:
- that logic would need to be very complex to catch e.g. `int ***q = &p; uintptr_t i = 0; memcpy(q, &i, sizeof(int**));`
- that logic would rarely trigger in real-world code, because a programmer usually declares a parameter as `int **p` instead of `int * const *p` when he or she wants to assign something to `*p`.

There are many situations where invalidations are questionable or even buggy, but I think this is one of the cases where the analyzer follows a reasonable behavior and still produces a false positive. This bug may be eliminated if we introduce a general purpose suppression for bugs where the analyzer assumes something (e.g. "p is null in the ternary operator") that contradicts the knowledge about the before-invalidation contents -- but there are no more specific solutions that would target this bug.

**Comment by bozicrHT** [comment]
Thanks a lot for the explanation. I understand that this example is not suitable for a small, targeted fix. Would it make sense to keep the issue open as a test case for the proposed general-purpose suppression, or should it be closed as "working as intended / known false positive"?
